Plug-n-play languages

DavidLegg opened the pull request:
DRAFT - using this PR to run CI, not yet ready to merge

cartermak committed:
Factor out language package from SequenceEditor

Commit:
Fix svelte type-checking errors

Commit:
Bug fix for loading sequence editor

Fixes a bug that prevented loading the sequence editor.
Don't try to dispatch changes to the editorSequenceView until we've initialized the editorSequenceView.

Commit:
Lock Playwright version to 1.49.1

Commit:
Remove extra dot in downloaded sequences

Commit:
Debug downloading sequences

Commit:
Clean up SequenceTemplates

Applies similar edits to SequenceTemplates & SequenceTemplateEditor as were applied to Sequences and SequenceEditor.
These factor (some) language-specifics out to a separate package, loaded through the parcel.

Notably, we decided to demand a "template language adaptation", rather than attempt to automatically produce a template adaptation from the regular language adaptation.
This gives end users the ability to apply different functionality to templates and regular sequences, since templating may interfere with some sequence language adaptation features.

This commit does not remove the "language" field from sequence templates, since that will require back end changes.
It does remove virtually all the functionality tied to that field, though.

Commit:
Fully remove ISequenceAdaptation

Fully removes the last traces of ISequenceAdaptation, along with a handful of related utility functions.

Commit:
Add getLibrarySequences to language adaptation

Add the function to detect library sequences to the language adaptation.
This removes the last hard-coded switch between VML and SeqN.

cartermak committed:
Notes

duranb commented on package.json:
```diff
@@ -151,4 +152,4 @@
     }
   }
-}
+}
```
Do you mind undoing any whitespace changes like this? This ending newline is automatically inserted via prettier.

duranb commented on src/app.d.ts:
```diff
@@ -102,4 +103,5 @@ declare global {
 }
 
-export {};
+export { };
```
Same as above. Unnecessary whitespace is automatically removed for objects.

duranb commented on src/components/sequencing/SequenceEditor.svelte:
```diff
@@ -74,19 +41,16 @@
-  export let sequenceDefinition: string = '';
+  export let newSequenceAdaptation: NewAdaptationInterface;
   export let sequenceName: string = '';
+  export let sequenceDefinition: string = ''; // TODO what on earth does this do
```
This is so that we can keep the original file text to use to compare with the `updatedSequenceDefinition` text (which is updated when the user types in the editor). If they're different, we know that the file has been modified.

duranb commented on src/language-package/languages/seq-n/output-linter.ts:
```diff
@@ -86,3 +74,21 @@ export function seqJsonLinter(view: EditorView, commandDictionary: CommandDictio
+ * @see https://github.com/codemirror/lang-json/blob/main/src/lint.ts
+ */
+function getErrorPosition(error: SyntaxError, doc: Text): number {
+  let m;
```
Do you mind renaming to `match` so it's intuitive as to what the variable's intent is for. Also, ideally, this should have a type as well.

duranb commented on src/language-package/languages/seq-n/output-linter.ts:
```diff
@@ -86,3 +74,21 @@ export function seqJsonLinter(view: EditorView, commandDictionary: CommandDictio
+function getErrorPosition(error: SyntaxError, doc: Text): number {
+  let m;
+
+  if ((m = error.message.match(/at position (\d+)/))) {
```
Do you mind avoiding inline setting the variable and evaluating of the conditional? I know it's valid, but it makes the code just a little more difficult to read. It's easy to miss that `m` is being set in the conditional and not evaluated as a comparison instead.

duranb commented on src/language-package/interfaces/new-adaptation-interface.ts:
```diff
@@ -0,0 +1,42 @@
+import type { Extension } from '@codemirror/state';
```
Is the intent to rename this file later on to not reflect the "new" status?

cartermak committed:
Delete language package

cartermak committed:
Update imports to retrieve from aerie-sequence-languages

cartermak committed:
Remove unused components

cartermak committed:
Remove unused dictionary now that VML lives elsewhere

cartermak committed:
Sequencing adaptation interface cleanup

Corresponds with an aerie-sequence-languages PR that does a lot of the final cleanup of the interface. Notably:

- 'adaptation' is now an interface with a getter for dependency injection that returns 'languages'. These are what are actually accessed by the editor largely.
- aerie-ui is no longer responsible for separately maintaining the library sequence list and map. Instead, adaptations may manipulate the array of library sequences.
- Via that dependency injection method, this repo now provides a utility to instantiate the editor tooltip Svelte components to keep that UI element concern separate from the aerie-sequence-languages repo.

cartermak commented:
Closing in favor of newer branch
